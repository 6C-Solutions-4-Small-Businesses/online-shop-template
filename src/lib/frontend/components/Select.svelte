<script lang="ts">
    import SelectOption from '$lib/frontend/components/SelectOption.svelte'
    import {writable, type Writable} from 'svelte/store'
    import ChevronDownIcon from '~icons/mdi/chevron-down'
    import ChevronUpIcon from '~icons/mdi/chevron-up'
    import Input from '$lib/frontend/components/Input.svelte'

    export let id: string
    export let label: string
    export let placeholder: string = 'Placeholder'
    export let selectedValue: string
    export let options: string []
    export let disabled: boolean = false
    export let required: boolean = false
    export let optionsBottom: boolean = true

    export let onChange: (value: string) => void = () => {
    }

    let showOptions: Writable<boolean> = writable(false)

    $:optionsListSize = options.length > 4 ? 'h-56' : 'h-auto'

    function toggleOptionsDisplay(): void {
        showOptions.set(!$showOptions)
    }

    function selectOption(value: string): void {
        onChange(value)
        showOptions.set(false)
    }

</script>

<div>
    <div class="relative">
        <Input
                disabled={disabled}
                id={id}
                label={label}
                placeholder={placeholder}
                value={selectedValue}
                required={required}
        />
        <button class="absolute bottom-2.5 right-1" on:click={toggleOptionsDisplay}>
            {#if $showOptions}
                <ChevronUpIcon class="w-8 h-8 text-gray-400"/>
            {:else}
                <ChevronDownIcon class="w-8 h-8 text-gray-400"/>
            {/if}
        </button>
        <div
                class={`${$showOptions ? 'visible bg-slate-100 shadow-lg p-2 rounded-lg overflow-scroll' : 'hidden'} absolute ${optionsBottom ? 'top-22' : 'bottom-14'} flex flex-col w-full font-bold gap-1 z-30 ${optionsListSize}`}>
            {#each options as option}
                <SelectOption
                        label={option}
                        isSelected={option === selectedValue}
                        onClickHandler={() => selectOption(option)}
                />
            {/each}
        </div>
    </div>
</div>
