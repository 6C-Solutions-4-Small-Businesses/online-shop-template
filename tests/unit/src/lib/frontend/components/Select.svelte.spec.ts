import {render, type RenderResult, screen} from '@testing-library/svelte'
import '@testing-library/jest-dom'
import Select from '$lib/frontend/components/Select.svelte'
import {expect} from 'vitest'

describe('Select Component', () => {
    const inputContainerId = 'test-id'
    const selectLabel = 'select-label'
    const selectPlaceHolder = 'input-placeholder'
    const initialSelectValue = 'Québec'
    const selectOptions = ['Québec', 'Ontario', 'British Columbia', 'Alberta', 'Manitoba', 'Saskatchewan', 'Nova Scotia', 'New Brunswick', 'Newfoundland and Labrador', 'Prince Edward Island', 'Northwest Territories', 'Yukon', 'Nunavut']
    const inputType = 'text'

    let rendering: RenderResult<Select, typeof import('@testing-library/dom/types/queries')>

    describe('html structure', () => {

        beforeEach(() => {
            rendering = render(Select, {
                props: {
                    id: inputContainerId,
                    label: selectLabel,
                    placeholder: selectPlaceHolder,
                    selectedValue: initialSelectValue,
                    options: selectOptions
                },
            })
        })

        it('should have match snapshot', () => {
            expect(rendering.container).toMatchSnapshot()
        })

        it('should map the "id" property to the "data-testid" of the containing div', () => {

            expect(screen.getByTestId(inputContainerId)).toBeInTheDocument()
        })

        it('should set the value of the label element to the value of the label property', () => {

            const labelElement = rendering.getByText(selectLabel)

            expect(labelElement).toBeInTheDocument()
        })

        it('should set the value of the input type to the value of the type property', () => {

            const inputElement = rendering.getByTestId(inputContainerId) as HTMLInputElement

            expect(inputElement.type).toBeDefined()
            expect(inputElement.type).toStrictEqual(inputType)
        })

        it('should set the value of the input placeholder to the value of the placeholder property', () => {

            const inputElement = rendering.getByTestId(inputContainerId) as HTMLInputElement

            expect(inputElement.placeholder).toBeDefined()
            expect(inputElement.placeholder).toStrictEqual(selectPlaceHolder)
        })

        it('should set the value of the input element to the initial value of the value property', () => {

            const inputElement = rendering.getByTestId(inputContainerId) as HTMLInputElement

            expect(inputElement.value).toBeDefined()
            expect(inputElement.value).toStrictEqual(initialSelectValue)
        })

        it('should set the input disabled property to the value of the provided disabled property', () => {

            const inputElement = rendering.getByTestId(inputContainerId) as HTMLInputElement

            expect(inputElement.disabled).toBeDefined()
            expect(inputElement.disabled).toBeFalsy()
        })

        test.each(selectOptions)('should have displayed the options', (option) => {

            const htmlElement = rendering.getByText(option)

            expect(htmlElement).toBeDefined()
        })
    })
})