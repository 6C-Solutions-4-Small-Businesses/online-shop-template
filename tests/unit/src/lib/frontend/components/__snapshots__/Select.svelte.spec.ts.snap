// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Select Component > html structure > should have match snapshot 1`] = `
<body>
  <div>
    <div>
      <div
        class="relative"
      >
        <div
          class="w-full"
        >
          <h4
            class="mb-1 text-black"
            data-testid="input-label-id"
          >
            select-label
             
          </h4>
           
          <input
            class="w-full p-3.5 rounded-md border-gray-500 bg-white focus:outline-none focus:border-blue-500"
            data-testid="test-id"
            id="test-id"
            placeholder="input-placeholder"
            type="text"
          />
        </div>
        <!--&lt;Input&gt;-->
         
        <button
          class="absolute bottom-2.5 right-1"
        >
          <svg
            class="w-8 h-8 text-gray-400"
            height="1.2em"
            viewBox="0 0 24 24"
            width="1.2em"
          >
            <path
              d="M7.41 8.58L12 13.17l4.59-4.59L18 10l-6 6l-6-6z"
              fill="currentColor"
            />
          </svg>
          <!--&lt;Chevron-down&gt;-->
        </button>
         
        <div
          class="hidden absolute top-22 flex flex-col w-full font-bold gap-4 z-30 h-56"
        >
          <button
            class="h-14 flex justify-start items-center shadow-md pl-3.5 py-4 hover:bg-orange-100 hover:text-red-500 rounded-md bg-orange-100 text-red-500"
          >
            Qu√©bec
          </button>
          <!--&lt;SelectOption&gt;-->
          <button
            class="h-14 flex justify-start items-center shadow-md pl-3.5 py-4 hover:bg-orange-100 hover:text-red-500 rounded-md bg-white"
          >
            Ontario
          </button>
          <!--&lt;SelectOption&gt;-->
          <button
            class="h-14 flex justify-start items-center shadow-md pl-3.5 py-4 hover:bg-orange-100 hover:text-red-500 rounded-md bg-white"
          >
            British Columbia
          </button>
          <!--&lt;SelectOption&gt;-->
          <button
            class="h-14 flex justify-start items-center shadow-md pl-3.5 py-4 hover:bg-orange-100 hover:text-red-500 rounded-md bg-white"
          >
            Alberta
          </button>
          <!--&lt;SelectOption&gt;-->
          <button
            class="h-14 flex justify-start items-center shadow-md pl-3.5 py-4 hover:bg-orange-100 hover:text-red-500 rounded-md bg-white"
          >
            Manitoba
          </button>
          <!--&lt;SelectOption&gt;-->
          <button
            class="h-14 flex justify-start items-center shadow-md pl-3.5 py-4 hover:bg-orange-100 hover:text-red-500 rounded-md bg-white"
          >
            Saskatchewan
          </button>
          <!--&lt;SelectOption&gt;-->
          <button
            class="h-14 flex justify-start items-center shadow-md pl-3.5 py-4 hover:bg-orange-100 hover:text-red-500 rounded-md bg-white"
          >
            Nova Scotia
          </button>
          <!--&lt;SelectOption&gt;-->
          <button
            class="h-14 flex justify-start items-center shadow-md pl-3.5 py-4 hover:bg-orange-100 hover:text-red-500 rounded-md bg-white"
          >
            New Brunswick
          </button>
          <!--&lt;SelectOption&gt;-->
          <button
            class="h-14 flex justify-start items-center shadow-md pl-3.5 py-4 hover:bg-orange-100 hover:text-red-500 rounded-md bg-white"
          >
            Newfoundland and Labrador
          </button>
          <!--&lt;SelectOption&gt;-->
          <button
            class="h-14 flex justify-start items-center shadow-md pl-3.5 py-4 hover:bg-orange-100 hover:text-red-500 rounded-md bg-white"
          >
            Prince Edward Island
          </button>
          <!--&lt;SelectOption&gt;-->
          <button
            class="h-14 flex justify-start items-center shadow-md pl-3.5 py-4 hover:bg-orange-100 hover:text-red-500 rounded-md bg-white"
          >
            Northwest Territories
          </button>
          <!--&lt;SelectOption&gt;-->
          <button
            class="h-14 flex justify-start items-center shadow-md pl-3.5 py-4 hover:bg-orange-100 hover:text-red-500 rounded-md bg-white"
          >
            Yukon
          </button>
          <!--&lt;SelectOption&gt;-->
          <button
            class="h-14 flex justify-start items-center shadow-md pl-3.5 py-4 hover:bg-orange-100 hover:text-red-500 rounded-md bg-white"
          >
            Nunavut
          </button>
          <!--&lt;SelectOption&gt;-->
        </div>
      </div>
    </div>
    <!--&lt;Select&gt;-->
  </div>
</body>
`;
